### PR TITLE
Events feed

### LINKED ISSUE #2392: Publish an RSS feed for Events
Similar to the OCaml Changelog, Events should be provided as a RSS feed (so that we can connect them with social media accounts on Twitter, LinkedIn, and Mastodon).

Hints:
- Look at the implementation of RSS feeds for the OCaml Changelog (https://github.com/ocaml/ocaml.org/blob/main/tool/ood-gen/lib/changelog.ml) as inspiration for how to do this for Events (https://github.com/ocaml/ocaml.org/blob/main/tool/ood-gen/lib/event.ml).
- The `create_feed` function for Events needs to be registered as a sub-command of `ood-gen` here: https://github.com/ocaml/ocaml.org/blob/main/tool/ood-gen/bin/gen.ml
- You need to add a dune rule at https://github.com/ocaml/ocaml.org/blob/main/dune that runs the `ood-gen` tool to create the `.xml` file.
- Update the template at `events.eml` to add a link to the RSS feed.

### PR DESCRIPTION
**Pull request by ishar19**
Resolves #2392

Creating a new PR with updated logic

### CODE REVIEW

**Comment by ishar19** [comment]
image

File: dune [hunk 1]
```diff
@@ -40,4 +40,18 @@
      %{target}
      (run %{gen_feed} changelog))))))
 
+(subdir
+ asset/
+ (rule
+  (target events.xml)
+  (deps
+   (source_tree %{workspace_root}/data/events)
+   (:ood_gen %{workspace_root}/tool/ood-gen/bin/feed.exe))
```
Comment (cuihtlauac) [inline dune]:
```suggestion
   (:gen_feed %{workspace_root}/tool/ood-gen/bin/feed.exe))
```

File: dune [hunk 2]
```diff
@@ -40,4 +40,18 @@
...
+  (action
+   (chdir
+    %{workspace_root}
+    (with-stdout-to
+     %{target}
+     (run %{ood_gen} events))))))
```
Comment (cuihtlauac) [inline dune]:
```suggestion
     (run %{gen_feed} events))))))
```

File: tool/ood-gen/lib/event.ml [hunk 3]
```diff
@@ -141,6 +141,47 @@ let all () =
...
+              let open Rss in
+              () |> all 
+              |> create_feed ~id:"events.xml" ~title:"OCaml Events" ~create_entry
+              |> feed_to_string
+              end
+        
+        
+        
+        
+
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion
```

File: tool/ood-gen/lib/event.ml [hunk 4]
```diff
@@ -141,6 +141,47 @@ let all () =
          in
          String.compare t2 t1)
 
+         module EventsFeed = struct
+          let create_entry (log : t) =
+            let authors = (Syndic.Atom.author "Ocaml.org", []) in
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion
            let authors = (Syndic.Atom.author "Ocaml.org", []) in
```

File: tool/ood-gen/lib/event.ml [hunk 5]
```diff
@@ -141,6 +141,47 @@ let all () =
...
+             let start_date_str = Syndic.Date.to_rfc3339 start_date in
+             let id = Uri.of_string (log.slug ^ " " ^ start_date_str) in
+             let location_summary =
+               match log.location with
+               | Some { lat; long } ->
+                   Printf.sprintf "%s (lat: %f, long: %f)" textual_location lat
+                     long
+               | None -> textual_location
+             in
+
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion
```

File: tool/ood-gen/lib/event.ml [hunk 6]
```diff
@@ -141,6 +141,47 @@ let all () =
...
+               ~title:(Syndic.Atom.Text (log.title ^ " " ^ start_date_str))
+               ~updated:start_date
+               ~links:[ Syndic.Atom.link (Uri.of_string log.url) ]
+               ~categories:[ Syndic.Atom.category event_type ]
+               ~content:(Syndic.Atom.Text location_summary) ()
+            
+      
+            let create_feed () =
+              let open Rss in
+              () |> all
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion
              () |> all
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]
Minor changes

File: tool/ood-gen/lib/event.ml [hunk 7]
```diff
@@ -141,6 +141,41 @@ let all () =
...
+               | None -> textual_location
+             in
+             Syndic.Atom.entry ~id ~authors
+               ~title:(Syndic.Atom.Text (log.title ^ " " ^ start_date_str))
+               ~updated:start_date
+               ~links:[ Syndic.Atom.link (Uri.of_string log.url) ]
+               ~categories:[ Syndic.Atom.category event_type ]
+               ~content:(Syndic.Atom.Text location_summary) ()
+            
+
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion

```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

**Comment by ishar19** [comment]
Sure thing, will update asap, AFK rn

File: tool/ood-gen/lib/event.ml [hunk 8]
```diff
@@ -141,6 +141,39 @@ let all () =
...
+               ~links:[ Syndic.Atom.link (Uri.of_string log.url) ]
+               ~categories:[ Syndic.Atom.category event_type ]
+               ~content:(Syndic.Atom.Text location_summary) ()
+            let create_feed () =
+              let open Rss in
+              () |> all
+              |> create_feed ~id:"events.xml" ~title:"OCaml Events" ~create_entry
+              |> feed_to_string
+              end
+
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion
  let create_entry (log : t) =
    let authors = (Syndic.Atom.author "Ocaml.org", []) in
    let event_type = EventType.show log.event_type in
    let textual_location = log.textual_location in
    let start_date =
      Syndic.Date.of_rfc3339
        (log.starts.yyyy_mm_dd ^ "T"
        ^ Option.value ~default:"00:00" log.starts.utc_hh_mm
        ^ ":00Z")
    in
    let start_date_str = Syndic.Date.to_rfc3339 start_date in
    let id = Uri.of_string (log.slug ^ " " ^ start_date_str) in
    let location_summary =
      match log.location with
      | Some { lat; long } ->
          Printf.sprintf "%s (lat: %f, long: %f)" textual_location lat long
      | None -> textual_location
    in
    Syndic.Atom.entry ~id ~authors
      ~title:(Syndic.Atom.Text (log.title ^ " " ^ start_date_str))
      ~updated:start_date
      ~links:[ Syndic.Atom.link (Uri.of_string log.url) ]
      ~categories:[ Syndic.Atom.category event_type ]
      ~content:(Syndic.Atom.Text location_summary) ()

  let create_feed () =
    let open Rss in
    () |> all
    |> create_feed ~id:"events.xml" ~title:"OCaml Events" ~create_entry
    |> feed_to_string
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]
Formatting

File: tool/ood-gen/lib/event.ml [hunk 9]
```diff
@@ -141,6 +141,36 @@ let all () =
...
+    let authors = (Syndic.Atom.author "Ocaml.org", []) in
+    let event_type = EventType.show log.event_type in
+    let textual_location = log.textual_location in
+    let start_date =
+      Syndic.Date.of_rfc3339
+        (log.starts.yyyy_mm_dd ^ "T"
+        ^ Option.value ~default:"00:00" log.starts.utc_hh_mm
+        ^ ":00Z")
+    in
+    let start_date_str = Syndic.Date.to_rfc3339 start_date in
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion
    let start_date_str = 
      log.starts.yyyy_mm_dd ^ "T"
      ^ Option.value ~default:"00:00" log.starts.utc_hh_mm
      ^ ":00Z"
```

Wouldn't it be the same this way?

File: tool/ood-gen/lib/event.ml [hunk 10]
```diff
@@ -141,6 +141,36 @@ let all () =
...
+    let id = Uri.of_string (log.slug ^ " " ^ start_date_str) in
+    let location_summary =
+      match log.location with
+      | Some { lat; long } ->
+          Printf.sprintf "%s (lat: %f, long: %f)" textual_location lat long
+      | None -> textual_location
+    in
+    Syndic.Atom.entry ~id ~authors
+      ~title:(Syndic.Atom.Text (log.title ^ " " ^ start_date_str))
+      ~updated:start_date
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml:109]: The event start date should not be used here. The updated date is when the event announcement is created or updated.

**Review by cuihtlauac — CHANGES_REQUESTED** [review]
Thanks @ishar19, this is nice. I've committed the trivial stuff. Two issues remain
* published_date is not set to the right value
* Date processing may have useless steps

File: tool/ood-gen/lib/event.ml (hunk 10 above)
Comment (ishar19) [reply tool/ood-gen/lib/event.ml:109]: Oh sure, I'm in my uni rn for an exam, will check back later

File: tool/ood-gen/lib/event.ml [hunk 11]
```diff
@@ -141,6 +141,36 @@ let all () =
          in
          String.compare t2 t1)
 
+  let create_entry (log : t) =
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml:86]:
```suggestion
  module EventsFeed = struct

  let create_entry (log : t) =
```

File: tool/ood-gen/lib/event.ml [hunk 12]
```diff
@@ -141,6 +141,36 @@ let all () =
...
+      ~updated:start_date
+      ~links:[ Syndic.Atom.link (Uri.of_string log.url) ]
+      ~categories:[ Syndic.Atom.category event_type ]
+      ~content:(Syndic.Atom.Text location_summary) ()
+
+  let create_feed () =
+    let open Rss in
+    () |> all
+    |> create_feed ~id:"events.xml" ~title:"OCaml Events" ~create_entry
+    |> feed_to_string
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml:118]:
```suggestion
    |> feed_to_string
    end
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

File: tool/ood-gen/lib/event.ml [hunk 13]
```diff
@@ -141,6 +141,39 @@ let all () =
          in
          String.compare t2 t1)
 
+  module EventsFeed = struct
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion
module EventsFeed = struct
```

File: tool/ood-gen/lib/event.ml [hunk 14]
```diff
@@ -141,6 +141,39 @@ let all () =
...
+      ~links:[ Syndic.Atom.link (Uri.of_string log.url) ]
+      ~categories:[ Syndic.Atom.category event_type ]
+      ~content:(Syndic.Atom.Text location_summary) ()
+
+  let create_feed () =
+    let open Rss in
+    () |> all
+    |> create_feed ~id:"events.xml" ~title:"OCaml Events" ~create_entry
+    |> feed_to_string
+    end
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion
end

```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

File: tool/ood-gen/lib/event.ml [hunk 15]
```diff
@@ -141,6 +141,40 @@ let all () =
          in
          String.compare t2 t1)
 
+module EventsFeed = struct
+
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]:
```suggestion
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

File: tool/ood-gen/lib/event.ml (hunk 10 above)
Comment (ishar19) [reply tool/ood-gen/lib/event.ml:109]: @cuihtlauac i checked the event feed markdown files, there isn't any updated date field
should i put a conditional whether when updated is present to use that or leave it as it is, thanks!

File: tool/ood-gen/lib/event.ml [hunk 16]
```diff
@@ -141,6 +141,36 @@ let all () =
...
+module EventsFeed = struct
+  let create_entry (log : t) =
+    let authors = (Syndic.Atom.author "Ocaml.org", []) in
+    let event_type = EventType.show log.event_type in
+    let textual_location = log.textual_location in
+    let start_date_str = 
+      log.starts.yyyy_mm_dd ^ "T"
+      ^ Option.value ~default:"00:00" log.starts.utc_hh_mm
+      ^ ":00Z"
+    let id = Uri.of_string (log.slug ^ " " ^ start_date_str) in
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/event.ml]: @ishar19 : please fix the syntax error here

File: tool/ood-gen/lib/event.ml (hunk 10 above)
Comment (cuihtlauac) [reply tool/ood-gen/lib/event.ml:109]: For 1-shot events, we need an additional field in `Event.t` to hold that information. When `all` is built, using the function `decode`, we must initialize it with the file's last write time on the markdown. 

For recurring events, it's trickier. We have to check in the RSS spec if it is allowed to skip that field.

**Review by cuihtlauac — DISMISSED** [review]

File: tool/ood-gen/lib/event.ml (hunk 10 above)
Comment (ishar19) [reply tool/ood-gen/lib/event.ml:109]: okay

Comment (ishar19) [reply tool/ood-gen/lib/event.ml:109]: I tried to research whether we can omit the field in recurring events or not but couldn't find anything substantial, will need some help here

Comment (cuihtlauac) [reply tool/ood-gen/lib/event.ml:109]: A possible solution could be to use the filesystem information for one-shot events and remove recurring events from the feed. Does it make sense to you?

**Comment by ishar19** [comment]
> A possible solution could be to use the filesystem information for one-shot events and remove recurring events from the feed. Does it make sense to you?

Can you please explain "filesystem information"
I get the removal of recurring events part

**Comment by cuihtlauac** [comment]
> Can you please explain "filesystem information" I get the removal of recurring events part

File modification information. Right now we're using `Bos.OS.File.read` to read the Yaml files. You could use `BOS.OS.Path.stat` to get the Unix.stats which contains the file last modification date

**Comment by ishar19** [comment]
> > Can you please explain "filesystem information" I get the removal of recurring events part
> 
> File modification information. Right now we're using `Bos.OS.File.read` to read the Yaml files. You could use `BOS.OS.Path.stat` to get the Unix.stats which contains the file last modification date

oh alright, thanks,i will look into it

**Review by sabine — APPROVED** [review]
Thanks so much for the contribution @ishar19, merging this now! :sparkles: :orange_heart:

**Pull request merged**